staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/status.go

```diff
@@ -58,26 +58,45 @@ func ErrorToAPIStatus(err error) *metav1.Status {
 		//TODO: check for invalid responses
 		return &status
 	default:
-		status := http.StatusInternalServerError
-		switch {
-		//TODO: replace me with NewConflictErr
-		case storage.IsConflict(err):
-			status = http.StatusConflict
-		}
 		// Log errors that were not converted to an error status
 		// by REST storage - these typically indicate programmer
 		// error by not using pkg/api/errors, or unexpected failure
 		// cases.
 		runtime.HandleError(fmt.Errorf("apiserver received an error that is not an metav1.Status: %#+v: %v", err, err))
-		return &metav1.Status{
-			TypeMeta: metav1.TypeMeta{
-				Kind:       "Status",
-				APIVersion: "v1",
-			},
-			Status:  metav1.StatusFailure,
-			Code:    int32(status),
-			Reason:  metav1.StatusReasonUnknown,
-			Message: err.Error(),
-		}
+		return convertStorageErrorOrNot(err)
```

Comment:
I think this is too late to do this conversion? There's been a coding logic error if we get here. We need to fix the logic error rather than paper over it.

Like I'm not sure how many different sources can produce these errors and if it is generally safe / helpful to convert back at this point, or if these specific conversions are always appropriate. What would definitely be helpful is to log enough information to help us diagnose where the error came from and why it hasn't already been converted.

Reply:
> We need to fix the logic error rather than paper over it.

There may be quite a few upstream logic errors, and finding them all is difficult. But eventually we need to find them all and fix them.

> if these specific conversions are always appropriate.

`convertStorageErrorOrNot` is a final line to try to convert storage error. Status, Code and Reason will be fine, but Message is not always appropriate, it's sad.

> What would definitely be helpful is to log enough information to help us diagnose where the error came from and why it hasn't already been converted.

That sounds great but I have no idea how to diagnose "where the error came from and why it hasn't already been converted."